src/lib/threadpool.c

```diff
@@ -0,0 +1,549 @@
+		ps_barrier:
+			break;
+		case PS_BARRIER:
+			if (!empty(u)) {
+				sm_move(planner_sm, PS_DRAINING_UNORD);
+				break;
+			}
+			if (pi->in_flight == 0) {
+				q = pop(o);
+				PRE(q_to_w(q)->type == WT_BAR);
+				free(q_to_w(q));
+				sm_move(planner_sm, PS_DRAINING);
+				break;
+			}
+			uv_cond_wait(&pi->planner_cond, mutex);
+			sm_move(planner_sm, PS_BARRIER);
+			break;
+		case PS_DRAINING_UNORD:
+			while (!empty(u)) {
+				q = pop(u);
+				push(&ts[q_tid(q)].inq, q);
+				uv_cond_signal(&ts[q_tid(q)].cond);
+			}
+			sm_move(planner_sm, PS_BARRIER);
+			break;
+		case PS_EXITED:
+			sm_fini(planner_sm);
+			uv_mutex_unlock(mutex);
+			return;
+		default:
+			POST(false && "Impossible!");
+		}
+	}
+}
+
+static void queue_work(pool_work_t *w)
+{
+	w->work_cb(w);
+}
+
+static void queue_done(pool_work_t *w)
+{
+	w_unregister(uv_loop_to_pool(w->loop), w);
+	if (w->after_work_cb != NULL)
+		w->after_work_cb(w);
+}
+
+static void worker(void *arg)
+{
+	struct targs        *ta = arg;
+	pool_impl_t         *pi = ta->pi;
+	uv_mutex_t          *mutex = &pi->mutex;
+	pool_thread_t       *ts = pi->threads;
+	enum pool_work_type  wtype;
+	pool_work_t         *w;
+	queue               *q;
+
+	uv_key_set(&pi->thread_key, &ta->idx);
+	uv_sem_post(ta->sem);
+	uv_mutex_lock(mutex);
+	for (;;) {
+		while (empty(&ts[ta->idx].inq)) {
+			if (pi->exiting) {
+				uv_mutex_unlock(mutex);
+				return;
+			}
+			uv_cond_wait(&ts[ta->idx].cond, mutex);
+		}
+
+		q = pop(&ts[ta->idx].inq);
+		uv_mutex_unlock(mutex);
+
+		w = q_to_w(q);
+		wtype = w->type;
+		queue_work(w);
+
+		uv_mutex_lock(&pi->outq_mutex);
+		push(&pi->outq, &w->link);
+		uv_async_send(&pi->outq_async);
+		uv_mutex_unlock(&pi->outq_mutex);
+
+		uv_mutex_lock(mutex);
+		if (wtype > WT_BAR) {
+			assert(pi->in_flight > 0);
+			if (--pi->in_flight == 0)
+				uv_cond_signal(&pi->planner_cond);
+		}
+	}
+}
+
+static void pool_cleanup(pool_t *loop)
+{
+	pool_impl_t   *pi = loop->pi;
+	pool_thread_t *ts = pi->threads;
+	uint32_t i;
+
+	if (pi->nthreads == 0)
+		return;
+
+	pi->exiting = true;
+	uv_cond_signal(&pi->planner_cond);
+
+	if (uv_thread_join(&pi->planner_thread))
+		abort();
+	uv_cond_destroy(&pi->planner_cond);
+	POST(empty(&pi->ordered) && empty(&pi->unordered));
+
+	for (i = 0; i < pi->nthreads; i++) {
+		uv_cond_signal(&ts[i].cond);
+		if (uv_thread_join(&ts[i].thread))
+			abort();
+		POST(empty(&ts[i].inq));
+		uv_cond_destroy(&ts[i].cond);
+	}
+
+	free(pi->threads);
+	uv_mutex_destroy(&pi->mutex);
+	uv_key_delete(&pi->thread_key);
+	pi->nthreads = 0;
+}
+
+static void pool_threads_init(pool_t *pool)
+{
+	uv_thread_options_t config;
+	const char *val;
+	uv_sem_t sem;
+	uint32_t i;
+	pool_impl_t *pi = pool->pi;
+	pool_thread_t *ts;
+	struct targs pa = (struct targs) {
+		.sem = &sem,
+		.pi = pi,
+	};
+
+	pi->qos = 0;
+	pi->o_prev = WT_BAR;
+	pi->exiting = false;
+	pi->in_flight = 0;
+	pi->nthreads = POOL_THREADPOOL_SIZE;
+
+	val = getenv("POOL_THREADPOOL_SIZE");
```

Comment:
Rather than getenv with a fallback in this file I'd prefer that the number of threads be set by an argument to `pool_init` -- the decision of how many threads to use should be made closer to the dqlite entry point